packages/@lwc/engine-dom/src/patches/global-registry.ts

```diff
@@ -0,0 +1,386 @@
+            }
+            return NativeCtor; // defined before the patched custom elements registry loaded
+        }
+    };
+
+    CustomElementRegistry.prototype.whenDefined = function whenDefined(
+        this: CustomElementRegistry,
+        tagName: string
+    ): Promise<CustomElementConstructor> {
+        return nativeWhenDefined.call(this, tagName).then(() => {
+            let promise = definedPromises.get(tagName);
+            if (isUndefined(promise)) {
+                const definition = definitionsByTag.get(tagName);
+                if (!isUndefined(definition)) {
+                    return definition.UserCtor;
+                }
+                let resolve: (constructor: CustomElementConstructor) => void;
+                promise = new Promise((r) => (resolve = r));
+                definedPromises.set(tagName, promise);
+                definedResolvers.set(tagName, (ctor) => resolve(ctor));
+            }
+            return promise;
+        });
+    };
+
+    function HTMLElement(this: HTMLElement) {
+        // Upgrading case: the pivoting class constructor was run by the browser's
+        // native custom elements and we're in the process of running the
+        // "constructor-call trick" on the natively constructed instance, so just
+        // return that here
+        const instance = upgradingInstance;
+        if (!isUndefined(instance)) {
+            upgradingInstance = undefined;
+            return instance;
+        }
+        // Construction case: we need to construct the pivoting instance and return it
+        // This is possible when the user register it via global registry and instantiate
+        // it via `new Ctor()`.
+        const { constructor } = this;
+        const definition = definitionsByClass.get(constructor as CustomElementConstructor);
+        if (isUndefined(definition) || !definition.PivotCtor) {
+            throw new TypeError('Illegal constructor');
+        }
+        // This constructor is ONLY invoked when it is the user instantiating
+        // an element via new Ctor while Ctor is a registered global constructor.
+        const { PivotCtor, UserCtor } = definition;
+        return new PivotCtor(UserCtor);
+    }
+    HTMLElement.prototype = NativeHTMLElement.prototype;
+    // @ts-ignore
+    window.HTMLElement = HTMLElement;
+
+    // this method patches the dom and returns helper function for LWC to register names and associate
+    // them to a constructor while returning the pivot constructor ready to instantiate via `new`
+    return (tagName: string, constructor: CustomElementConstructor): CustomElementConstructor => {
+        tagName = tagName.toLowerCase();
+        let PivotCtor = pivotCtorByTag.get(tagName);
+        if (isUndefined(PivotCtor)) {
+            const definition = getDefinitionForConstructor(constructor);
+            PivotCtor = createPivotingClass(tagName, definition);
+            definitionsByTag.set(tagName, definition);
```

Comment:
@nolanlawson I think this line is the main problem. IMO, definitionsByTag should be only used for globally defined names, and should never be populated by this LWC internal mechanism to register pivots. It might be more complicated than that, but I will take a look in details tomorrow.

Reply:
actually, by looking at the code... YES, this line MUST be removed to avoid polluting the global names.

Reply:
After poking around with the code for a bit, I think we may need both "local" definitions by class/tag as well as "global" definitions by class/tag.

There are definitely some odd edge cases here, though. I'm still writing tests to try to capture everything.